pyro/distributions/empirical.py

```diff
@@ -62,89 +44,47 @@ def sample_size(self):
 
         :return int: number of samples collected.
         """
-        self._finalize()
-        if self._samples is None:
-            return 0
-        return self._samples.size(0)
-
-    def add(self, value, weight=None, log_weight=None):
-        """
-        Adds a new data point to the sample. The values in successive calls to
-        ``add`` must have the same tensor shape and size. Optionally, an
-        importance weight can be specified via ``log_weight`` or ``weight``
-        (default value of `1` is used if not specified).
-
-        :param torch.Tensor value: tensor to add to the sample.
-        :param torch.Tensor weight: log weight (optional) corresponding
-            to the sample.
-        :param torch.Tensor log_weight: weight (optional) corresponding
-            to the sample.
-        """
-        if self._validate_args:
-            if weight is not None and log_weight is not None:
-                raise ValueError("Only one of ```weight`` or ``log_weight`` should be specified.")
-
-        weight_type = value.new_empty(1).float().type() if value.dtype in (torch.int32, torch.int64) \
-            else value.type()
-        # Apply default weight of 1.0.
-        if log_weight is None and weight is None:
-            log_weight = torch.tensor(0.0).type(weight_type)
-        elif weight is not None and log_weight is None:
-            log_weight = math.log(weight)
-        if isinstance(log_weight, numbers.Number):
-            log_weight = torch.tensor(log_weight).type(weight_type)
-        if self._validate_args and log_weight.dim() > 0:
-            raise ValueError("``weight.dim() > 0``, but weight should be a scalar.")
-
-        # Seed the container tensors with the correct tensor types
-        if self._samples is None:
-            self._samples = value.new_tensor([])
-            self._log_weights = log_weight.new_tensor([])
-        # Append to the buffer list
-        self._samples_buffer.append(value)
-        self._weights_buffer.append(log_weight)
+        return self._log_weights.numel()
 
     def sample(self, sample_shape=torch.Size()):
-        self._finalize()
-        idxs = self._categorical.sample(sample_shape=sample_shape)
-        return self._samples[idxs]
+        sample_idx = self._categorical.sample(sample_shape)
+        return self._samples[sample_idx]
 
     def log_prob(self, value):
         """
         Returns the log of the probability mass function evaluated at ``value``.
         Note that this currently only supports scoring values with empty
-        ``sample_shape``, i.e. an arbitrary batched sample is not allowed.
```

Comment:
@eb8680 - This PR is getting a bit unwieldy, but I will generalize this to an arbitrary `sample_shape` in a follow-up and add some tests. Currently samples are restricted to be of shape `batch_shape + event_shape` but it should be easy to have arbitrary batch dims on the left.